src/Tools/dotnet-monitor/EndpointInfo/ServerEndpointInfoSource.cs

```diff
@@ -248,33 +162,119 @@ private async Task ResumeAndQueueEndpointInfo(IpcEndpointInfo info, Cancellation
+
+                await PruneEndpointsAsync(validEndpoints: null, token);
             }
         }
 
-        private void VerifyIsListening()
+        private async Task NotifyAndRemoveAsync(ReversedDiagnosticsServer server, CancellationToken token)
         {
-            if (!IsListening)
+            while (!token.IsCancellationRequested)
             {
-                throw new InvalidOperationException(nameof(ServerEndpointInfoSource.Start) + " method must be called before invoking this operation.");
+                IEndpointInfo endpoint = await _pendingRemovalReader.ReadAsync(token);
+
+                foreach (IEndpointInfoSourceCallbacks callback in _callbacks)
+                {
+                    await callback.OnRemovedEndpointInfoAsync(endpoint, token).ConfigureAwait(false);
+                }
+
+                server.RemoveConnection(endpoint.RuntimeInstanceCookie);
             }
         }
 
-        private bool IsListening => null != _server && null != _listenTask;
+        private async Task PruneEndpointsAsync(List<IEndpointInfo> validEndpoints, CancellationToken token)
+        {
+            List<IEndpointInfo> endpointsToRemove = new();
+
+            // Prune connections that no longer have an active runtime instance before
+            // returning the list of connections.
+            await _activeEndpointsSemaphore.WaitAsync(token).ConfigureAwait(false);
+
+            try
+            {
+                // Check the transport for each endpoint info and remove it if the check fails.
+                List<Task<bool>> checkTasks = new();
+                foreach (EndpointInfo info in _activeEndpoints)
+                {
+                    checkTasks.Add(Task.Run(() => CheckEndpointAsync(info, token), token));
+                }
+
+                // Wait for all checks to complete
+                bool[] results = await Task.WhenAll(checkTasks).ConfigureAwait(false);
+
+                // Remove failed endpoints from active list; record the failed endpoints
+                // for removal after releasing the active endpoints semaphore.
+                int endpointIndex = 0;
+                for (int resultIndex = 0; resultIndex < results.Length; resultIndex++)
+                {
+                    IEndpointInfo endpoint = _activeEndpoints[endpointIndex];
+                    if (results[resultIndex])
+                    {
+                        validEndpoints?.Add(endpoint);
+                        endpointIndex++;
+                    }
+                    else
+                    {
+                        _activeEndpoints.RemoveAt(endpointIndex);
+
+                        endpointsToRemove.Add(endpoint);
+                    }
+                }
+            }
+            finally
+            {
+                _activeEndpointsSemaphore.Release();
+            }
+
+            if (endpointsToRemove.Count > 0)
+            {
+                foreach (IEndpointInfo endpoint in endpointsToRemove)
```

Comment:
We are pruning on a timer and every time we get all the endpoints. Maybe do the writing inside the semaphore to avoid multiple writers?